### DISCUSSION TITLE
Test function failure, not using vm.prank on purpose to understand the difference between msg.sender and address(this) better.

### DISCUSSION THREAD

**Discussion by sushrut-d07** [Cyfrin/foundry-full-course-cu#2514] [Q&A]
`Ran 7 tests for test/FundMeTest.t.sol:FundMeTest
[PASS] testAddsFunderToArrayOfFunders() (gas: 106717)
[PASS] testFundFailsWithoutEnoughEth() (gas: 30050)
[PASS] testFundUpdatesFundedDataStructure() (gas: 106450)
[PASS] testMinimumDollarIsFive() (gas: 8323)
[PASS] testOnlyOwnerCanWithdraw() (gas: 106176)
[PASS] testOwnderIsSender() (gas: 8487)
[FAIL: assertion failed: 79228162536908014944548096175 != 0] testWithdrawIsWorkingWithOwner() (gas: 117895)
Suite result: FAILED. 6 passed; 1 failed; 0 skipped; finished in 6.52s (11.54s CPU time)

Ran 1 test suite in 7.25s (6.52s CPU time): 6 tests passed, 1 failed, 0 skipped (7 total tests)

Failing tests:
Encountered 1 failing test in test/FundMeTest.t.sol:FundMeTest
[FAIL: assertion failed: 79228162536908014944548096175 != 0] testWithdrawIsWorkingWithOwner() (gas: 117895)

Encountered a total of 1 failing tests, 6 tests succeeded

THIS IS MY ENTIRE CODE
''' solidity
//SPDX-License-Identifier:Unlicensed

pragma solidity ^0.8.26;

import {Test, console} from "lib/forge-std/src/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
    FundMe fundMe;
    uint256 send_amount = 0.1 ether;

    function setUp() external {
        // fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    modifier funded() {
        fundMe.fund{value: send_amount}();
        _;
    }

    function testMinimumDollarIsFive() public view {
        assertEq(fundMe.minimumUsd(), 5e18);
    }

    function testOwnderIsSender() public view {
        // a test needs to start with test to be inclued in forge test by solidity
        assertEq(fundMe.getOwner(), msg.sender);
    }

    function testFundFailsWithoutEnoughEth() public {
        vm.expectRevert(); //This means that the next line should revert, if it doesnt revert the whole test fails
        fundMe.fund();
    }

    function testFundUpdatesFundedDataStructure() public funded {
        uint256 amountFunded = fundMe.getAddressToAmountFunded(address(this));
        assertEq(amountFunded, send_amount);
    }

    function testAddsFunderToArrayOfFunders() public funded {
        address sender = fundMe.getFunder(0);
        assertEq(sender, address(this));
    }

    function testOnlyOwnerCanWithdraw() public funded {
        vm.expectRevert();
        fundMe.withdraw();
    }

    function testWithdrawIsWorkingWithOwner() public funded {
        //Arrange
        uint256 startingOwnerBalance = fundMe.getOwner().balance;
        uint256 startingFundMeBalance = address(fundMe).balance;
        //Act
        vm.prank(fundMe.getOwner());
        fundMe.withdraw();
        //Assert
        uint256 endingOwnerBalance = fundMe.getOwner().balance;
        uint256 endingFundMeBalance = address(fundMe).balance;

        assertEq(startingOwnerBalance, 0);
        assertEq(
            startingFundMeBalance + endingFundMeBalance,
            endingOwnerBalance
        );
    }
}
'''

**Comment by EngrPips** [comment]
Hello @sushrut-d07, what is the issue here?

**Reply by sushrut-d07** [reply]
I am sorry for writing this ugly question, I'll frame it better here in this comment
.
I figured how to run the test but i deduced one thing from my solution that I dont clearly understand 
testWithdrawIsWorkingWithOwner() function in my code fails if i use "uint256 startingOwnerBalance = fundMe.getOwner().balance"
but with" address owner = fundMe.getOwner();
startingOwnerBalance = owner.balance;" it passes

**Reply by EngrPips** [reply]
Thats weird. Can you run the two processes with the `-vvvv` flag and share with us the stack trace of the test process?
